### PR TITLE
Resubmit of failed comments

### PR DESCRIPTION
**Pull request by sdorra**
## Proposed changes

Ports the resubmit feature from the jira plugin to the issue tracker plugin to make it available for every issue tracker. The feature includes resending of failed comments and notifications.

### Your checklist for this pull request

- [X] PR is well described and the description can be used as commit message on squash
- [X] Related issues linked to PR if existing and labels set
- [X] Target branch is not master (in most cases develop should bet the target of choice) 
- [X] Code does not conflict with target branch
- [X] New code is covered with unit tests
- [X] [Changelog entry file](https://github.com/scm-manager/changelog#changelog-entry-files) created in `gradle/changelog`
- [X] Definition of Done's fulfilled: [DoD](https://github.com/scm-manager/scm-manager/blob/develop/docs/en/development/definition-of-done.md) // [UI DoD](https://github.com/scm-manager/scm-manager/blob/develop/docs/en/development/ui-dod.md)
- [X] Documentation updated (only necessary for new features or changed behaviour)

### Checklist for branch merge request (not required for forks)

- [x] Branch is green/blue on [Jenkins](https://oss.cloudogu.com/jenkins/)
- [ ] Quality Gate passed on [SonarQube](https://sonarcloud.io/organizations/scm-manager/projects)

### CODE REVIEW

File: src/main/js/admin/ResubmitSection.tsx [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+            </td>
+            <td>
+              <Action
+                label={t("scm-issuetracker-plugin.resubmit.clear")}
+                linkName={"clear"}
+                resubmit={resubmit}
+                color="danger"
+                actionDispatcher={actionDispatcher}
+              />
+            </td>
```
Comment (pfeuffer) [inline src/main/js/admin/ResubmitSection.tsx:93]: The buttons are somewhat out of bounds (or the text in the other columns is not centered enough). Somehow this looks a bit out of order.

File: src/main/resources/locales/de/plugins.json [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+    "title": "Issue Tracker",
+    "subtitle": "Allgemeine Issue Tracker Konfiguration",
+    "navLink": "Issue Tracker",
+    "resubmit": {
+      "title": "Kommentare erneut senden",
+      "description": "Ausstehende Kommentare erneut senden",
+      "empty": "Keine ausstehenden Kommentare gefunden",
+      "issueTracker": "Issue Tracker",
+      "queueCount": "Anzahl ausstehender Kommentare",
+      "resubmit": "Erneute senden",
```
Comment (pfeuffer) [inline src/main/resources/locales/de/plugins.json:12]: Typo..?
```suggestion
      "resubmit": "Erneut senden",
```

File: src/main/resources/locales/de/plugins.json [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+      "clear": "Kommentare entfernen",
+      "config": {
+        "title": "Erneut Senden Konfiguration",
```
Comment (pfeuffer) [inline src/main/resources/locales/de/plugins.json:15]:
```suggestion
        "title": "Konfiguration für erneutes Senden",
```

File: src/main/resources/locales/de/plugins.json [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+        "description": "Konfiguration für die Erneute Senden Funktion",
```
Comment (pfeuffer) [inline src/main/resources/locales/de/plugins.json:16]:
```suggestion
        "description": "Konfiguration für die \"Erneut Senden\" Funktion",
```

File: src/main/resources/locales/de/plugins.json [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+        "updated": "Die Konfiguration wurde erfolgreich aktualisiert",
+        "submit": "Konfiguration aktualisieren"
+      }
+    }
+  },
+  "permissions": {
+    "issuetracker": {
+      "resubmit": {
+        "displayName": "Kommentare erneut senden",
+        "description": "Erneutes senden ausstehender Issue Tracker Kommentare veranlassen und konfigurieren"
```
Comment (pfeuffer) [inline src/main/resources/locales/de/plugins.json:30]:
```suggestion
        "description": "Erneutes Senden ausstehender Issue Tracker Kommentare veranlassen und konfigurieren"
```

File: src/main/resources/locales/en/plugins.json [hunk 6]
```diff
@@ -0,0 +1,34 @@
+{
+  "scm-issuetracker-plugin": {
+    "title": "Issue Tracker",
+    "subtitle": "General Issue Tracker configuration",
```
Comment (pfeuffer) [inline src/main/resources/locales/en/plugins.json:4]: Either all caps or not.
```suggestion
    "subtitle": "General issue tracker configuration",
```

File: src/main/resources/sonia/scm/issuetracker/internal/resubmit/comment-mail.md.mustache [hunk 7]
```diff
@@ -0,0 +1,9 @@
+# Failed to create comment on {{ comment.issueTracker }} for issue {{ comment.issueKey }}
+
+Failed to create the following comment on issue tracker {{ comment.issueTracker }}:
+
+```code
+{{{ comment.comment }}}
+```
+
+The comment can be sent again in the [issue tracker admin section]({{ url }}).
```
Comment (pfeuffer) [inline src/main/resources/sonia/scm/issuetracker/internal/resubmit/comment-mail.md.mustache:9]: Don't we have a german version for this?

File: src/main/resources/sonia/scm/issuetracker/internal/resubmit/resubmit-mail.md.mustache [hunk 8]
```diff
@@ -0,0 +1,6 @@
+# Resubmit for issue tracker {{ issueTracker }}
+
+* **{{ removeCount }}** successfully resubmitted
+* **{{ requeueCount }}** could not be added and are requeued
+
+Requeued comments can be resend or removed in the [issue tracker admin section]({{ url }}).
```
Comment (pfeuffer) [inline src/main/resources/sonia/scm/issuetracker/internal/resubmit/resubmit-mail.md.mustache:6]: Don't we have a german version for this?

File: src/test/java/sonia/scm/issuetracker/internal/resubmit/EvictingQueueTest.java [hunk 9]
```diff
@@ -0,0 +1,49 @@
...
+    assertThat(queue.add("a")).isFalse();
+  }
+
+  @Test
+  void shouldEvictFirstAddedEntry() {
+    Queue<String> queue = EvictingQueue.create(2);
+    assertThat(queue.add("a")).isTrue();
+    assertThat(queue.add("b")).isTrue();
+    assertThat(queue.add("c")).isTrue();
+    assertThat(queue).containsOnly("b", "c");
```
Comment (pfeuffer) [inline src/test/java/sonia/scm/issuetracker/internal/resubmit/EvictingQueueTest.java:46]: Had to think twice to distinguish action and checked result here. Think there are two tests hidden inside :-)

File: src/test/java/sonia/scm/issuetracker/internal/resubmit/ResubmitConfigurationStoreTest.java [hunk 10]
```diff
@@ -0,0 +1,76 @@
...
+
+  @Test
+  @SubjectAware("trillian")
+  void shouldFailWithoutPermission() {
+    ResubmitConfiguration configuration = new ResubmitConfiguration();
+    assertThrows(UnauthorizedException.class, () -> store.set(configuration));
+  }
+
+  @Test
+  @SubjectAware(value = "trillian", permissions = "issuetracker:resubmit:*")
```
Comment (pfeuffer) [inline src/test/java/sonia/scm/issuetracker/internal/resubmit/ResubmitConfigurationStoreTest.java:57]: Like this new annotation 👍🏽

**Review by pfeuffer — CHANGES_REQUESTED** [review]
Minor remarks only 👍🏽

File: src/main/java/sonia/scm/issuetracker/internal/resubmit/ResubmitProcessor.java [hunk 11]
```diff
@@ -0,0 +1,137 @@
...
+    }
+  }
+
+  private void process(IssueTracker tracker, Collection<QueuedComment> queuedComments) {
+    Optional<Resubmitter> resubmitter = tracker.getResubmitter();
+    if (resubmitter.isPresent()) {
+      process(resubmitter.get(), queuedComments);
+    } else {
+      LOG.warn("tracker {} does not support resubmit, requeue {} comments", tracker.getName(), queuedComments.size());
+      requeue.addAll(queuedComments);
```
[the remark this answers is not in the document]
Comment (sdorra) [reply src/main/java/sonia/scm/issuetracker/internal/resubmit/ResubmitProcessor.java:103]: A the moment we store every failed comment, even if the issue tracker does not support resubmitting.

File: src/main/java/sonia/scm/issuetracker/internal/resubmit/ResubmitResource.java [hunk 12]
```diff
@@ -0,0 +1,246 @@
...
+  @ApiResponse(responseCode = "403", description = "not authorized, the current user does not have the \"resubmit\" privilege")
+  @ApiResponse(
+    responseCode = "500",
+    description = "internal server error",
+    content = @Content(
+      mediaType = VndMediaType.ERROR_TYPE,
+      schema = @Schema(implementation = ErrorDto.class)
+    )
+  )
+  public ResubmitConfigurationDto getConfiguration(@Context UriInfo info) {
```
[the remark this answers is not in the document]
Comment (sdorra) [reply src/main/java/sonia/scm/issuetracker/internal/resubmit/ResubmitResource.java:102]: The name is only used if the `@Operation` has no `operationId`.

File: src/main/java/sonia/scm/issuetracker/internal/resubmit/ResubmitResource.java [hunk 13]
```diff
@@ -0,0 +1,246 @@
...
+  @ApiResponse(responseCode = "403", description = "not authorized, the current user does not have the \"resubmit\" privilege")
+  @ApiResponse(
+    responseCode = "500",
+    description = "internal server error",
+    content = @Content(
+      mediaType = VndMediaType.ERROR_TYPE,
+      schema = @Schema(implementation = ErrorDto.class)
+    )
+  )
+  public Response setConfiguration(@Valid ResubmitConfigurationDto dto) {
```
[the remark this answers is not in the document]
Comment (sdorra) [reply src/main/java/sonia/scm/issuetracker/internal/resubmit/ResubmitResource.java:134]: The name is only used if the `@Operation` has no `operationId`.

File: src/main/java/sonia/scm/issuetracker/internal/resubmit/ResubmitResource.java [hunk 14]
```diff
@@ -0,0 +1,246 @@
...
+  @ApiResponse(responseCode = "403", description = "not authorized, the current user does not have the \"resubmit\" privilege")
+  @ApiResponse(
+    responseCode = "500",
+    description = "internal server error",
+    content = @Content(
+      mediaType = VndMediaType.ERROR_TYPE,
+      schema = @Schema(implementation = ErrorDto.class)
+    )
+  )
+  public Response resubmit(@PathParam("issueTracker") String issueTracker) {
```
[the remark this answers is not in the document]
Comment (sdorra) [reply src/main/java/sonia/scm/issuetracker/internal/resubmit/ResubmitResource.java:217]: The name is only used if the `@Operation` has no `operationId`.

File: src/main/java/sonia/scm/issuetracker/internal/resubmit/ResubmitResource.java [hunk 15]
```diff
@@ -0,0 +1,246 @@
...
+  @ApiResponse(responseCode = "403", description = "not authorized, the current user does not have the \"resubmit\" privilege")
+  @ApiResponse(
+    responseCode = "500",
+    description = "internal server error",
+    content = @Content(
+      mediaType = VndMediaType.ERROR_TYPE,
+      schema = @Schema(implementation = ErrorDto.class)
+    )
+  )
+  public Response clear(@PathParam("issueTracker") String issueTracker) {
```
[the remark this answers is not in the document]
Comment (sdorra) [reply src/main/java/sonia/scm/issuetracker/internal/resubmit/ResubmitResource.java:241]: The name is only used if the `@Operation` has no `operationId`.

File: src/main/java/sonia/scm/issuetracker/internal/resubmit/ResubmitProcessor.java (hunk 11 above)
[the remark this answers is not in the document]
Comment (pfeuffer) [reply src/main/java/sonia/scm/issuetracker/internal/resubmit/ResubmitProcessor.java:103]: Is it wise to requeue them in this case? This sounds like there is no chance to get rid of them and they would have to be dumped sooner or later.

File: src/main/java/sonia/scm/issuetracker/internal/resubmit/ResubmitResource.java (hunk 12 above)
[the remark this answers is not in the document]
Comment (pfeuffer) [reply src/main/java/sonia/scm/issuetracker/internal/resubmit/ResubmitResource.java:102]: Good tip, thx 👍🏽

File: src/main/java/sonia/scm/issuetracker/internal/resubmit/ResubmitConfigurationStore.java [hunk 16]
```diff
@@ -0,0 +1,55 @@
...
+
+  private final ConfigurationStore<ResubmitConfiguration> store;
+
+  @Inject
+  public ResubmitConfigurationStore(ConfigurationStoreFactory storeFactory) {
+    this.store = storeFactory.withType(ResubmitConfiguration.class).withName(STORE_NAME).build();
+  }
+
+  public ResubmitConfiguration get() {
+    return store.getOptional().orElse(new ResubmitConfiguration());
```
Comment (pfeuffer) [inline src/main/java/sonia/scm/issuetracker/internal/resubmit/ResubmitConfigurationStore.java:47]:
```suggestion
    return store.getOptional().orElseGet(() -> new ResubmitConfiguration());
```

File: src/main/js/admin/ResubmitSection.tsx [hunk 17]
```diff
@@ -0,0 +1,123 @@
...
+          <th>{t("scm-issuetracker-plugin.resubmit.queueCount")}</th>
+          <th />
+          <th />
+        </tr>
+      </thead>
+      <tbody>
+        {data?._embedded.resubmit?.map(resubmit => (
+          <tr className="border-is-green">
+            <td>{resubmit.issueTracker}</td>
+            <td>{resubmit.queueSize}</td>
```
Comment (pfeuffer) [inline src/main/js/admin/ResubmitSection.tsx:80]: These cells should be centered vertically to align them with the buttons

File: src/main/resources/locales/de/plugins.json [hunk 18]
```diff
@@ -0,0 +1,34 @@
...
+      "clear": "Kommentare entfernen",
+      "config": {
+        "title": "Konfiguration für erneutes Senden",
+        "description": "Konfiguration für die \"Erneut Senden\" Funktion",
+        "address": "Email Adresse",
+        "addAddress": "Adresse hinzufügen",
+        "removeAddress": "Adresse entfernen",
+        "invalidMail": "Falsches Email Adressformat",
+        "updated": "Die Konfiguration wurde erfolgreich aktualisiert",
+        "submit": "Konfiguration aktualisieren"
```
Comment (pfeuffer) [inline src/main/resources/locales/de/plugins.json:22]:
```suggestion
        "submit": "Speichern"
```

File: src/main/resources/locales/en/plugins.json [hunk 19]
```diff
@@ -0,0 +1,34 @@
...
+      "clear": "Clear",
+      "config": {
+        "title": "Resubmit configuration",
+        "description": "Configuration related to the resubmit function",
+        "address": "Email Address",
+        "addAddress": "Add address",
+        "removeAddress": "Remove address",
+        "invalidMail": "Invalid email address format",
+        "updated": "Successfully updated resubmit configuration",
+        "submit": "Update configuration"
```
Comment (pfeuffer) [inline src/main/resources/locales/en/plugins.json:22]:
```suggestion
        "submit": "Submit"
```

File: src/main/resources/sonia/scm/issuetracker/internal/resubmit/resubmit-mail_de.mustache [hunk 20]
```diff
@@ -0,0 +1,7 @@
+# Erneutes Senden der Komemntare für Issue Tracker {{ issueTracker }}
```
Comment (pfeuffer) [inline src/main/resources/sonia/scm/issuetracker/internal/resubmit/resubmit-mail_de.mustache:1]:
```suggestion
# Erneutes Senden der Kommentare für Issue Tracker {{ issueTracker }}
```

File: src/main/resources/sonia/scm/issuetracker/internal/resubmit/resubmit-mail_de.mustache [hunk 21]
```diff
@@ -0,0 +1,7 @@
...
+
+* **{{ removeCount }}** erfolgreich gesendet
+* **{{ requeueCount }}** konnten nicht gesendet werden und werden wieder eingereiht
+
+Wieder eingereihte Kommentare können über die
+[Issue Tracker Admin Oberfläche]({{ url }}) erneute gesendet oder gelöscht werden.
```
Comment (pfeuffer) [inline src/main/resources/sonia/scm/issuetracker/internal/resubmit/resubmit-mail_de.mustache:7]:
```suggestion
[Issue Tracker Admin Oberfläche]({{ url }}) erneut gesendet oder gelöscht werden.
```

**Review by pfeuffer — CHANGES_REQUESTED** [review]
Small last remarks

**Pull request merged**